mlflow/pyfunc/__init__.py

```diff
@@ -201,20 +203,62 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
 
     :param spark: A SparkSession object.
     :param path: A path containing a :py:mod:`mlflow.pyfunc` model.
     :param run_id: ID of the run that produced this model. If provided, ``run_id`` is used to
                    retrieve the model logged with MLflow.
-    :return: Spark UDF type returned by the model's prediction method. Default double.
+    :param result_type: the return type of the user-defined function. The value can be either a
+                        :class:`pyspark.sql.types.DataType` object or a DDL-formatted type string.
+                        Only a primitive type or an array of primitive types are allowed.
+                        The following classes of result type are supported:
+                        - "int" or instanceof IntegerType: The leftmost integer that can fit in
+                          int32 result is returned or exception is raised if there is none.
+                        - "long" or instanceof LongType: The leftmost long integer that can fit in
+                           int64 result is returned or exception is raised if there is none.
+                        - ArrayType(IntegerType|LongType): Return all integer columns that can fit
+                          into the requested size.
+                        - "float" or instanceof FloatType: The leftmost numeric result cast to
+                          float32 is returned or exception is raised if there is no numeric column.
+                        - "double" or instanceof DoubleType: The leftmost numeric result cast to
+                          double is returned or exception is raised if there is no numeric column.
+                        - ArrayType(FloatType|DoubleType): Return all numeric columns cast to the
+                          requested. type. Exception is raised if there are numeric columns.
+                        - "string" or StringType: Result is the leftmost column converted to string.
+                        - Array[StringType]: Return all columns converted to string.
+
+    :return: Spark UDF which will apply model's prediction method to the data. Default double.
     """
 
     # Scope Spark import to this method so users don't need pyspark to use non-Spark-related
     # functionality.
     from mlflow.pyfunc.spark_model_cache import SparkModelCache
     from pyspark.sql.functions import pandas_udf
+    from pyspark.sql.types import _parse_datatype_string
+    from pyspark.sql.types import ArrayType, DataType
+    from pyspark.sql.types import DoubleType, IntegerType, FloatType, LongType, StringType
+
+    if not isinstance(result_type, DataType):
+        result_type = _parse_datatype_string(result_type)
+
+    elem_type = result_type
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+
+    supported_types = [IntegerType, LongType, FloatType, DoubleType, StringType]
+
+    if sum([1 if isinstance(elem_type, x) else 0 for x in supported_types]) == 0:
```

Comment:
This might be more readable as `if not any([isinstance(elem_type, x) for x in supported_types])`